Index: src\core\scroll.c
===================================================================
--- src\core\scroll.c
+++ src\core\scroll.c
@@ -12,8 +12,9 @@
 #include "math.h"
 #include "fade_manager.h"
 #include "parallax.h"
 #include "palette.h"
+#include "events.h"
 
 // put submap of a large map to screen
 void set_bkg_submap(UINT8 x, UINT8 y, UINT8 w, UINT8 h, const unsigned char *map, UINT8 map_w) OLDCALL;
 
@@ -44,8 +45,35 @@
 INT16 current_col, new_col;
 
 static FASTUBYTE _save_bank;
 
+script_event_t render_events[3];
+
+void assign_render_script(SCRIPT_CTX * THIS) OLDCALL BANKED {
+    UWORD *slot = VM_REF_TO_PTR(FN_ARG2);
+    UBYTE *bank = VM_REF_TO_PTR(FN_ARG1);
+    UBYTE **ptr = VM_REF_TO_PTR(FN_ARG0);
+    render_events[*slot].script_bank = *bank;
+    render_events[*slot].script_addr = *ptr;
+}
+
+void clear_render_script(SCRIPT_CTX * THIS) OLDCALL BANKED {
+    UWORD *slot = VM_REF_TO_PTR(FN_ARG0);
+    render_events[*slot].script_bank = 0;
+    render_events[*slot].script_addr = NULL;
+}
+
+static void scroll_callback_execute(UBYTE i)
+{
+    script_event_t *event = &render_events[i];
+    if (!event->script_addr)
+        return;
+    if ((event->handle == 0) || ((event->handle & SCRIPT_TERMINATED) != 0))
+    {
+        script_execute(event->script_bank, event->script_addr, &event->handle, 0, 0);
+    }
+}
+
 void scroll_init(void) BANKED {
     draw_scroll_x   = 0;
     draw_scroll_y   = 0;
     scroll_x_min    = 0;
@@ -62,9 +90,9 @@
     pending_h_i     = 0;
     scroll_x        = 0x7FFF;
     scroll_y        = 0x7FFF;
 
-    game_time       = 0; // was in scroll_render_rows() - that is insane, here is not the best place either 
+    game_time       = 0; // was in scroll_render_rows() - that is insane, here is not the best place either
 }
 
 void scroll_update(void) BANKED {
     INT16 x, y;
@@ -110,25 +138,28 @@
         } else {
             shift_scroll_x = draw_scroll_x >> port->shift;
         }
 
-        port->shadow_scx = shift_scroll_x;        
+        port->shadow_scx = shift_scroll_x;
         UBYTE shift_col = PX_TO_TILE(shift_scroll_x);
 
         // If column is +/- 1 just render next column
         if (current_col == new_col - 1) {
             // Render right column
             UBYTE x = shift_col - SCREEN_PAD_LEFT + SCREEN_TILE_REFRES_W - 1;
             scroll_load_col(x, port->start_tile, port->tile_height);
+            scroll_callback_execute(0);
         } else if (current_col == new_col + 1) {
             // Render left column
             UBYTE x = MAX(0, shift_col - SCREEN_PAD_LEFT);
             scroll_load_col(x, port->start_tile, port->tile_height);
+            scroll_callback_execute(0);
         } else if (current_col != new_col) {
             // If column differs by more than 1 render entire viewport
             scroll_render_rows(shift_scroll_x, 0, port->start_tile, port->tile_height);
-        }  
-        return FALSE;   
+            scroll_callback_execute(2);
+        }
+        return FALSE;
     } else {
         // last parallax slice OR no parallax
         port->shadow_scx = draw_scroll_x;
 
@@ -149,11 +180,13 @@
             activate_actors_in_col(x, full_y);
         } else if (current_col != new_col) {
             // If column differs by more than 1 render entire screen
             scroll_render_rows(draw_scroll_x, draw_scroll_y, ((scene_LCD_type == LCD_parallax) ? port->start_tile : -SCREEN_PAD_TOP), SCREEN_TILE_REFRES_H);
+            scroll_callback_execute(2);
             return TRUE;
         } else if (pending_h_i) {
             scroll_load_pending_col();
+            scroll_callback_execute(0);
         }
 
         // If row is +/- 1 just render next row
         if (current_row == new_row - 1) {
@@ -170,11 +203,13 @@
             activate_actors_in_row(x, y);
         } else if (current_row != new_row) {
             // If row differs by more than 1 render entire screen
             scroll_render_rows(draw_scroll_x, draw_scroll_y, ((scene_LCD_type == LCD_parallax) ? port->start_tile : -SCREEN_PAD_TOP), SCREEN_TILE_REFRES_H);
+            scroll_callback_execute(2);
             return TRUE;
         } else if (pending_w_i) {
             scroll_load_pending_row();
+            scroll_callback_execute(1);
         }
 
         return TRUE;
     }
@@ -201,10 +236,11 @@
 
 void scroll_queue_row(UBYTE x, UBYTE y) {
     while (pending_w_i) {
         // If previous row wasn't fully rendered
-        // render it now before starting next row        
+        // render it now before starting next row
         scroll_load_pending_row();
+        scroll_callback_execute(1);
     }
 
     // Don't queue rows past image height
     if (y >= image_tile_height) {
@@ -220,8 +256,9 @@
     while (pending_h_i) {
         // If previous column wasn't fully rendered
         // render it now before starting next column
         scroll_load_pending_col();
+        scroll_callback_execute(0);
     }
 
     pending_h_x = x;
     pending_h_y = y;
@@ -271,9 +308,9 @@
 }
 
 void scroll_load_col(UBYTE x, UBYTE y, UBYTE height) NONBANKED {
     _save_bank = CURRENT_BANK;
- 
+
 #ifdef CGB
     if (_is_CGB) {  // Color Column Load
         SWITCH_ROM(image_attr_bank);
         VBK_REG = 1;
@@ -307,5 +344,6 @@
     pending_h_y += height;
     pending_h_i -= height;
 
     SWITCH_ROM(_save_bank);
+
 }
